R/types-chr.R

```diff
@@ -0,0 +1,295 @@
+#' Construct a character vector or a string.
+#'
+#' These base-type constructors allow more control over the creation
+#' of strings in R. They take character vectors or string-like objects
+#' (integerish or raw vectors), and optionally set the encoding. The
+#' string version checks that the input contains a scalar string.
+#'
+#' @param x A character vector or a vector or list of string-like
+#'   objects.
+#' @param encoding If non-null, passed to
+#'   \code{\link{chr_set_encoding}()} to add an encoding mark.
+#' @seealso \code{\link{chr_set_encoding}()} for more information
+#'   about encodings in R.
+#' @export
+#' @examples
+#' # As everywhere in R, you can specify a string with Unicode
+#' # escapes. The characters corresponding to Unicode codepoints will
+#' # be encoded in UTF-8, and the string will be marked as UTF-8
+#' # automatically:
+#' cafe <- string("caf\uE9")
+#' str_encoding(cafe)
+#' str_bytes(cafe)
+#'
+#' # In addition, string() provides useful conversions to let
+#' # programmers control how the string is represented in memory. For
+#' # encodings other than UTF-8, you'll need to supply the bytes in
+#' # hexadecimal form. If it is a latin1 encoding, you can mark the
+#' # string explicitly:
+#' cafe_latin1 <- string(c(0x63, 0x61, 0x66, 0xE9), "latin1")
+#' str_encoding(cafe_latin1)
+#' str_bytes(cafe_latin1)
+#'
+#' # chr() accepts lists and will apply string() to each element:
+#' chr(list(cafe, c(0x63, 0x61, 0x66, 0xE9)))
+string <- function(x, encoding = NULL) {
+  if (is_integerish(x)) {
+    x <- rawToChar(as.raw(x))
+  } else if (is_raw(x)) {
+    x <- rawToChar(x)
+  } else if (!is_string(x)) {
+    abort("`x` must be a string or raw vector")
+  }
+
+  if (!is_null(encoding)) {
+    x <- chr_set_encoding(x, encoding)
```

Comment:
Maybe make `chr_set_encoding` do nothing is `encoding` is null? Would save a bit of duplication here